Replace IdentityServer4 with OpenIddict and upgrade to .net 10

juliangiebel opened the pull request:
**!!!This PR was locally tested extensively. It should still be tested with different oauth clients used in production though. Esp. the wiki as it can potentially break those oauth clients/application. Ideally upgrading SS14.Web is tested on a separate instance created from a snapshot first!!!**

This PR updates the project to .net 9 and replaces IdentityServer4 with OpenIddict while trying to ensure that existing OAuth applications keep working. For that purpose there is a data migration sql script under `tools/identityserver4_to_openiddict_data_migration.sql`.

## Differences between IdentityServer4 and OpenIddict
The biggest difference between the two OIDC solutions is the way client/applications are stored and the amount of extra features and settings.
OpenIddict uses a *lot* less database entities to store applications as it stores most non queried settings and configurations as json inside the application entity. It also doesn't normalize entities to quite the extent IS4 does.

This means there wasn't a need for a separate schema. 
Also a lot of settings that weren't used got removed from the UI as implementing them with OpenIddict would mean adding custom settings and implementing custom event handlers (Which this PR already does for settings that had to be re-implemented like  "PlainPkce" and "Allow PS256")

Encryption and Signing keys have to be supplied as certificates using PFX files.

OpenIddict also uses a different hashing algorithm for secrets and it salts them so migrated keys from IS4 are marked as legacy and handled seperatly using a re-implementation of the the way secrets are hashed in IS4 (Which uses standard .net cryptography methods).

## Additional changes
- With the upgrade to .net 9 this PR also consolidates `Program.cs` and `Startup.cs` for `SS14.Web/SS14.Web` into a single top lever `Program.cs` file.
- Some files use `#nullable true`.
- Custom implementations for postgres type conversions/handlers weren't necessary anymore and got removed.

## Noteworthy improvements
- The `/.well-known/openid-configuration` now only lists actually supported capabilities and features
- Access and authentication tokens are encrypted (Which might've also been already the case with IS4)
- Requested scopes are now present in the ID token in addition to getting them the `userinfo` endpoint 

## Generating encryption and signing certificates
At least one encryption and one signing certificate needs to be generated and configured for each encryption/signing algorithm that needs to be supported.
For this purpose a csx file exists under `tools\GenerateCerts.csx`(I used the .net10 preview to execute it otherwise it needs a proper project file).
Executing that file will generate 3 certificates. One encryption certificate using RS265 and two signing certificates using RS265 and PS265 respectively.

### Example certificate configuration
```yml
OpenId:
  Certificates:
    DefaultSigningAlgorithm: RS256
    EncryptionCertificates:
      - Path: 'C:\Users\julia\Projects\SS14.Web\cert\server-encryption-certificate.pfx'
    SigningCertificates:
      - Path: 'C:\Users\julia\Projects\SS14.Web\cert\server-signing-certificate-rsa-pss.pfx'
        Algorithm: PS256
      - Path: 'C:\Users\julia\Projects\SS14.Web\cert\server-signing-certificate-rsa.pfx'
```

Fixes #20  
Fixes #17

juliangiebel committed:
Remove IdentityServer4 packages

juliangiebel committed:
Comment out IdentityServer4 code

juliangiebel committed:
Upgrade to .net9
Upgrade nuget packages
Fix npgsql errors from upgrade
Note: (IpAddress, int) is now NpgsqlCidr

juliangiebel committed:
Refactor SS14.Web startup to program.cs

juliangiebel committed:
Work on adding OpenIddict to program.cs

juliangiebel committed:
Implement OpenIddict configuration

juliangiebel committed:
Implement custom application entity
Migrate to nullable

juliangiebel committed:
Revert "migrate to nullable"

This reverts commit 7b0520a9f95fc67b04458a77b6a69b0943ed9283.

juliangiebel committed:
Add list of ef core commands to readme

juliangiebel committed:
A

juliangiebel committed:
Fix wrong get openiddict request method call

juliangiebel committed:
Finish implementing the oidc authorization/consent endpoint

juliangiebel committed:
Clean up Consent.cshtml and codebehind

juliangiebel committed:
Remove IS4 code from ApplicationDbContext

juliangiebel committed:
Remove UserOauthClient

juliangiebel committed:
Change PersonalDataCollector to collecto from OpenIddict

juliangiebel committed:
Implement generate token event handler for ensuring the right algorithm is used

juliangiebel committed:
Add home page url property to SpaceApplication

juliangiebel committed:
Work on oauth application manage page

juliangiebel committed:
Implement multiple client secret support

juliangiebel committed:
Make a migration

juliangiebel committed:
Implement methods for working with multiple secrets
Implement unit tests for multiple secret handling methods

juliangiebel committed:
Re-Implement developer and oauth manage page

juliangiebel committed:
Re-Implement oauth application create page

juliangiebel committed:
Re-Implement confirm delete page
Implement custom url validator

juliangiebel committed:
Implement identityserver4 legacy SHA256 client secret validation

juliangiebel committed:
Re-Implement oauth client admin index and delete pages

juliangiebel committed:
Implement plain pkce option
Work on admin oauth app settings page

juliangiebel committed:
Re-Implement oauth client admin edit page

juliangiebel committed:
Add quartz for openiddict pruning
Implement legacy token handling test

juliangiebel committed:
Work on IdentityServer4Migration

juliangiebel committed:
Improve certificate configuration
Fix token lifetime settings

juliangiebel committed:
Remove test data seeder

juliangiebel committed:
Add data migration sql to tools

juliangiebel committed:
Address leftover todos and comments

VasilisThePikachu commented on Tools/identityserver4_to_openiddict_data_migration.sql:
Lowercase SQL is a crime

VasilisThePikachu reviewed (changes requested):
Ok well one thing you SHOULD actually do it update the dotnet version for the test so it wont fail

Also i guess look into net 10 support now, OpenIddict does support it

juliangiebel committed:
Revert "Remove test data seeder"

This reverts commit 9fd9eec76c80ed3aaa8a51352b14deaa88d57d13.

juliangiebel committed:
Update to .net10

juliangiebel committed:
Update dependencies

juliangiebel committed:
Set openiddict logging to warning outside of development

juliangiebel committed:
Add configuration setting for enabling test data seeding

juliangiebel committed:
Remove unimplemented and unused methods

juliangiebel committed:
Add documentation comments for OpenIdActionService

juliangiebel committed:
Add rewrite rule for connect/authorize

This is necessary to keep the endpoint path that's been used previously while having the authorize logic inside the consent page.
That avoids having to haul around the auth request data between redirects (which would be annoying to implement at this point)

Fildrance commented on SS14.Web/PersonalDataCollector.cs:
```diff
@@ -1,24 +1,32 @@
 using Microsoft.Extensions.Logging;
+using OpenIddict.Core;
 using SS14.Auth.Shared.Data;
+using SS14.Web.Extensions;
```
unused using

Fildrance commented on SS14.Web/PersonalDataCollector.cs:
```diff
@@ -128,52 +137,68 @@ private async Task CollectPatrons(ZipArchive zip, SpaceUser user, CancellationTo
+            Scopes: x.Scopes
+            ));
 
-        StringToFile(zip, "UserOAuthClients.json", data);
```
now it is unused but it is good method to have i think, so let it be?

Fildrance commented on SS14.Web/OpenId/Services/SignedInIdentityService.cs:
```diff
@@ -0,0 +1,43 @@
+{
+    private readonly IHttpContextAccessor _httpContextAccessor;
+    private readonly UserManager<SpaceUser> _userManager;
+    private readonly SignInManager<SpaceUser> _signInManager;
```
unused field

Fildrance commented on SS14.Web/OpenId/Services/SignedInIdentityService.cs:
```diff
@@ -0,0 +1,43 @@
+        if (_httpContextAccessor.HttpContext is not { } context)
+            return null;
+
+        var user = await _userManager.GetUserAsync(context.User);
```
Its a little bit sad that we have to do 2 calls for storage - GetUserAsync and GetUserIdAsync, but yeah, i don't see any nicer way... sadge

Fildrance commented on SS14.Web/OpenId/Services/IdentityClaimsProvider.cs:
```diff
@@ -0,0 +1,67 @@
+    /// <summary>
+    /// Provides claims for the specified identity.
+    /// </summary>
+    /// <param name="userId"></param>
```
don't leave them empty :godo:

Fildrance commented on SS14.Web/OpenId/Services/IdentityClaimsProvider.cs:
```diff
@@ -0,0 +1,67 @@
+        if (user == null)
+            return;
+
+        var username = await _userManager.GetUserNameAsync(user) ?? "";
```
better use String.Empty instead of "" because that clarifies that there were no mistake and empty string IS the intent.

Fildrance commented on SS14.Web/OpenId/Services/IdentityClaimsProvider.cs:
```diff
@@ -0,0 +1,67 @@
+            identity.AddClaim(new Claim(Claims.Email, await _userManager.GetEmailAsync(user) ?? ""));
+
+        if (scopes.Contains(OpenIddictConstants.Scopes.Roles))
+            identity.AddClaims(Claims.Role, ImmutableArray.Create([.. await _userManager.GetRolesAsync(user)]));
```
this method looks like 4 round-trips to DB in worst scenario - won't it be kinda slow? :(

Fildrance commented on SS14.Web/OpenId/Configuration/CertificateOptions.cs:
```diff
@@ -0,0 +1,9 @@
+
+public sealed class CertificateOptions
+{
+    public string Path { get; set; } = null!;
```
xml-doc for this class would not hurt, specifically fact that password and algorithm are mandatory (and maybe enum of recommended alorithms. You can also make validation for options on-boot, instead of making fields nullable :>

Fildrance commented on SS14.Web/OpenId/Configuration/OpenIdCertificateConfiguration.cs:
```diff
@@ -0,0 +1,30 @@
+
+public sealed class OpenIdCertificateConfiguration
+{
+    public const string Name = "Certificates";
```
its a bit inconsistent on the whole codebase - names of configuration sections, some have Name const inside them, some are named Options, others - Configuration. I can fix it after we merge this, as long as PR works, it just cleanup...

Fildrance commented on SS14.Web/OpenId/Configuration/OpenIdCertificateConfiguration.cs:
```diff
@@ -0,0 +1,30 @@
+    /// <remarks>
+    /// A certificate with the matching algorithm needs to be present
+    /// </remarks>
+    public string? DefaultEncryptionAlgorithm { get; set; }
```
Usually Default thing is used when nested options (customization) is not specified (and it cannot be null), but there it works inverted (default can be null and then it uses customization options) - i guess that is intentional, but i am not getting why really.

Fildrance commented on SS14.Web/OpenId/OpenIdConstants.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using System.Collections.Generic;
```
unused usings

Fildrance commented on SS14.Web/OpenId/Types/AuthorizationResult.cs:
```diff
@@ -0,0 +1,40 @@
+    ClaimsPrincipal? Principal)
+{
+    public static AuthorizationResult SignIn(ClaimsPrincipal principal) =>
+        new(ResultType.SignIn, null, null, null, principal);
```
Inconsistent mehod body declaration style even in one class

Fildrance commented on SS14.Web/OpenId/Types/AuthenticationValidationFailure.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public static AuthenticationValidationFailure LoginRequired => new(Error: OpenIddictConstants.Errors.LoginRequired);
+
+    public static AuthenticationValidationFailure Challenge (AuthenticationProperties properties) =>
```
No point in extra whitespae between method name and its arguments

Fildrance commented on SS14.Web/OpenId/Types/ConsentResult.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace SS14.Web.OpenId.Types;
+
+public record ConsentResult(ConsentResult.ResultType Type, string? ErrorName, ClaimsPrincipal? Principal)
```
if you want to enforce rules for object creation, you should also make c-tor private. This will also make intellisense use you static methods instead of proposing usage of c-tor.  (applicable to all such classes in PR)

Fildrance commented on SS14.Web/Models/SpaceUrlAttribute.cs:
```diff
@@ -0,0 +1,18 @@
+namespace SS14.Web.Models;
+
+[AttributeUsage(AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Parameter)]
+public class SpaceUrlAttribute : ValidationAttribute
```
I guess you want it to provide ability for setting something more then just https/http? Better write that into xml-doc then :|

Fildrance commented on SS14.Web/OpenId/EventHandlers/TokenSigningHandler.cs:
```diff
@@ -0,0 +1,81 @@
+
+        var encryptionAlgorithm = settings.TryGetValue(OpenIdConstants.EncryptionAlgorithmSetting, out var encryptionAlg)
+            ? encryptionAlg
+            : _config.DefaultEncryptionAlgorithm;
```
you are making _config.DefaultEncryptionAlgorithm assign on next line, so this probably is not really doing anything?

Fildrance commented on SS14.Web/OpenId/EventHandlers/TokenSigningHandler.cs:
```diff
@@ -0,0 +1,81 @@
+        {
+            foreach (var credential in context.Options.EncryptionCredentials)
+            {
+                if (!credential.Enc.Equals(encryptionAlgorithm))
```
its no gamecode so you should not be so afraid to use LINQ - since First is actually going to fit there perfectly, but your choice ofc

same applies to signing credentials

Fildrance commented on SS14.Web/OpenId/EventHandlers/TokenSigningHandler.cs:
```diff
@@ -0,0 +1,81 @@
+
+        var signingAlgorithm = settings.TryGetValue(OpenIdConstants.SigningAlgorithmSetting, out var signingAlg)
+            ? signingAlg
+            : _config.DefaultSigningAlgorithm;
```
same with double assignment of default

Fildrance commented on SS14.Web/OpenId/Extensions/ApplicationManagerExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+
+    public static async ValueTask<List<SpaceApplication>> FindApplicationsByUserId(this OpenIddictApplicationManager<SpaceApplication> manager, Guid userId, CancellationToken cancel = default)
+    {
+        return await manager.ListAsync(x => x.Where(a => a.SpaceUserId == userId)
+            , cancel).ToListAsync(cancellationToken: cancel);
```
```suggestion
        return await manager.ListAsync(
            x => x.Where(a => a.SpaceUserId == userId),
            cancel
        ).ToListAsync(cancellationToken: cancel);
```
:slugwhat:

Fildrance commented on SS14.Web/OpenId/Extensions/OpenIdExtension.cs:
```diff
@@ -0,0 +1,99 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.IdentityModel.Tokens;
+using OpenIddict.Abstractions;
```
unused using

Fildrance commented on SS14.Web/OpenId/Extensions/OpenIdExtension.cs:
```diff
@@ -0,0 +1,99 @@
+
+namespace SS14.Web.OpenId.Extensions;
+
+public static class OpenIdExtension
```
usually DI related stuff is cluttered in some specific top-level folder, and here it is kinda scattered between logical domains... a little bit of a meh. not critical i guess

Fildrance commented on SS14.Web/Models/Types/Result.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace SS14.Web.Models.Types;
+
+public record Result<TResult, TError>(TResult? Value, TError? Error) where TResult : class where TError : class
```
again, private c-tor would ensure you won't get broken object. And record have 'with' syntax so it wont work that way also

Fildrance commented on SS14.Web/OpenId/Services/OpenIdActionService.cs:
```diff
@@ -0,0 +1,220 @@
+namespace SS14.Web.OpenId.Services;
+
+/**
+ * Handles all the OpenIddict action plumbing. Used inside the consent page.
```
why not normal xml-doc?

Fildrance commented on SS14.Web/OpenId/Services/SpaceApplicationManager.cs:
```diff
@@ -0,0 +1,228 @@
+            secret = secret[5..];
+
+        var secrets = secret.Split(',');
+        var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString();
```
no need for ToString, interpolation will call it by itself

Fildrance commented on SS14.Web/OpenId/Services/SpaceApplicationManager.cs:
```diff
@@ -0,0 +1,228 @@
+        string comparand,
+        CancellationToken ct = new())
+    {
+        var secrets = comparand.Split(',');
```
writing down template for secret somewhere in this class as a comment would be quite nice for understanding all string related manipulatins :godo:

Fildrance commented on SS14.Web/OpenId/Services/SpaceApplicationManager.cs:
```diff
@@ -0,0 +1,228 @@
+            var description = desc ?? obfuscatedSecret[^6..];
+            var secretInfoString = $"{id}.{timestamp}.{obfuscatedSecret}.{description}";
+            var secretsString = secrets is null ? secretInfoString : $"{secrets},{secretInfoString}";
+            var secretInfo = new ClientSecretInfo(id, DateTimeOffset.UtcNow, description, false);
```
better take it into variable and use once

Fildrance commented on SS14.Web/OpenId/Services/SpaceApplicationManager.cs:
```diff
@@ -0,0 +1,228 @@
+            return (secretsString, secretInfo);
+        }
+
+        public static string? RemoveSecret(string secrets, int id)
```
this stuff works and covers all required use cases, but my god its hard to read and understand due to span-ification :( basically 'DO NOT TOUCH THIS' territory and not tests... on the other hand - this is basically deep internal code, so its probably okay-ish

Fildrance commented on SS14.Web/Extensions/PatreonExtension.cs:
```diff
@@ -0,0 +1,43 @@
+                options.Events.OnCreatingTicket += context =>
+                {
+                    var handler = context.HttpContext.RequestServices.GetService<PatreonConnectionHandler>();
+                    return handler!.HookCreatingTicket(context);
```
if you are so sure it wont be null - use GetRequiredService to get it

Fildrance commented on SS14.Web/OpenId/TestDataSeeder.cs:
```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Security.Cryptography;
```
unused using

Fildrance commented on SS14.Web/OpenId/TestDataSeeder.cs:
```diff
@@ -0,0 +1,109 @@
+
+namespace SS14.Web.OpenId;
+
+public sealed class TestDataSeeder(IServiceProvider serviceProvider) : IHostedService
```
is there a point to not use BackgrounService as parent? Are you sure need to block app start until this completes?

Fildrance commented on SS14.Web/OpenId/TestDataSeeder.cs:
```diff
@@ -0,0 +1,109 @@
+public sealed class TestDataSeeder(IServiceProvider serviceProvider) : IHostedService
+{
+    public async Task StartAsync(CancellationToken ct)
+    {
```
please wrap the whole thing in exception logging :3

Fildrance commented on SS14.Web/OpenId/TestDataSeeder.cs:
```diff
@@ -0,0 +1,109 @@
+
+        // Logging throws an invalid operation exception if an account gets created by something like an I Hosted service
+        // for seeding data. Throwing here doesn't prevent creating the account so it gets logged as an error instead
+        // TODO: Ask if I can make the GetActorIP method log an error instead of throwing instead
```
it is nullable, so i see no reason for throwing (except maybe rename it to TryGet or Find?

Fildrance commented on SS14.Web/OpenId/TestDataSeeder.cs:
```diff
@@ -0,0 +1,109 @@
+        };
+
+        // Logging throws an invalid operation exception if an account gets created by something like an I Hosted service
+        // for seeding data. Throwing here doesn't prevent creating the account so it gets logged as an error instead
```
Was not getting that exception - maybe related to concurrency of a kind? Also we should wrap this whole thing in logging, so it should work for hosted services for sure :(

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Manage/OAuthApps/Create.cshtml.cs:
```diff
@@ -1,43 +1,51 @@
 
+        //[Url(ErrorMessage = "The Homepage URL field is not a valid fully-qualified http or https URL.")]
         [Required]
+        [SpaceUrl]
```
other attribute is commented out and ErrorMessage is not used?

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Manage/OAuthApps/Create.cshtml.cs:
```diff
@@ -1,43 +1,51 @@
+    private readonly SpaceApplicationManager _appManager;
 
-    [BindProperty] public InputModel Input { get; set; }
+    [BindProperty] public InputModel Input { get; set; }  = null!;
```
some excessive double whitespaces in this file...

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Manage/OAuthApps/Create.cshtml.cs:
```diff
@@ -46,42 +54,42 @@ public void OnGet()
+            DisplayName = Input.Name,
+            RedirectUris = { new Uri(Input.CallbackUrl) },
+            Requirements = { OpenIddictConstants.Requirements.Features.ProofKeyForCodeExchange },
+            Settings = {{OpenIdConstants.AllowPlainPkce, "true"}},
```
probably needs a little bit of whitespaces

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Manage/OAuthApps/Create.cshtml.cs:
```diff
@@ -46,42 +54,42 @@ public void OnGet()
-
-        await _dbContext.SaveChangesAsync();
+        var app = await _appManager.CreateAsync(appDescriptor);
+        app.SpaceUserId = user!.Id;
```
Q_Q i know thats not really needed, but maybe check for null and return not found at start? null suppressions are triggering...

Fildrance commented on SS14.Web/Areas/Admin/Pages/Clients/Index.cshtml:
```diff
@@ -30,19 +30,19 @@
+                    @if (app.SpaceUserId is { } id)
                     {
-                        <a asp-page="../Users/ViewUser" asp-route-id="@user.Id">@user.UserName</a>
+                        <a asp-page="../Users/ViewUser" asp-route-id="@id">@(await Model.GetUserNameAsync(id))</a>
```
kind reminder that this one does not use cache and if we will have 20+ of those - we will have 20+ consequtive calls for storage, i believe.

Fildrance commented on SS14.Web/Areas/Admin/Pages/Clients/Index.cshtml.cs:
```diff
@@ -1,50 +1,73 @@
-            select new { c, uc };
-            
-        Clients = (await query.ToListAsync()).Select(c => (c.c, c.uc));
+        Apps = await _appManager.ListAsync().ToListAsync();
```
Propagating CancellationToken is supported (just add as an argument), and should be probably used where appropriate

Fildrance commented on SS14.Web/Areas/Admin/Pages/Clients/Client.cshtml.cs:
```diff
@@ -1,325 +1,246 @@
+        return RedirectToPage(new {id});
+    }
+
+    public async Task<IActionResult> OnPostDeleteSecretAsync(string id, int secret)
```
if we open 'create new client' form and then click delete - it will crash :(

Fildrance commented on SS14.Web/Areas/Admin/Pages/Clients/ConfirmDelete.cshtml:
```diff
@@ -9,11 +9,11 @@
-    <button id="delete-button" type="submit" asp-page-handler="Delete" asp-route-id="@Model.DbClient.Id" class="btn btn-danger">Delete</button>
-</form>
\ No newline at end of file
+    <button id="delete-button" type="submit" asp-page-handler="Delete" asp-route-id="@Model.App?.Id" class="btn btn-danger">Delete</button>
```
might be nice to add something like 
```
Please confirm that you want to delete client @Model.Title
<br>
```
because this page could appear quite confusing.
ideally add creator display name, but this would require more code so meh.

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Consent.cshtml:
```diff
@@ -1,10 +1,13 @@
 ﻿@page
+@using Microsoft.Extensions.Primitives
+@using OpenIddict.Abstractions
+@using SS14.Web.Extensions
```
unused using

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Consent.cshtml:
```diff
@@ -19,66 +22,69 @@
         <div class="col d-flex align-items-center">
-            <h2 class="text-center flex-grow-1 m-3">Authorize @Model.AuthRequest.Client.ClientName</h2>
-            @if (!string.IsNullOrWhiteSpace(@Model.AuthRequest.Client.LogoUri))
+            <h2 class="text-center flex-grow-1 m-3">Authorize @*@Model.AuthRequest.Client.ClientName*@</h2>
```
why is it just sitting there commented out?

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Consent.cshtml:
```diff
@@ -19,66 +22,69 @@
+            @if (!string.IsNullOrWhiteSpace(Model.Application?.LogoUri))
             {
-                <img class="rounded" height="48" src="@Model.AuthRequest.Client.LogoUri" alt="@Model.AuthRequest.Client.ClientName"/>
+                <img class="rounded" height="48" src="@Model.Application?.LogoUri" alt="@Model.Application?.DisplayName"/>
```
it does not seem like there is a way to set up LogoUri - is this for future use?

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Consent.cshtml.cs:
```diff
@@ -1,83 +1,141 @@
+    {
+        return Input?.Button switch
+        {
+            "no" => Forbid(OpenIddictServerAspNetCoreDefaults.AuthenticationScheme),
```
maybe put those into constants and use them in view too? also i think its possible to use true/false and parse it, but i am guessing there is some formality that needs to be accounted.

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Consent.cshtml.cs:
```diff
@@ -1,83 +1,141 @@
         {
-            response.Error = AuthorizationError.AccessDenied;
+            case false when validation.Error.IsChallenge:
+                return Challenge(validation.Error.Properties!);
```
i got to this view, then logged out on other window, then i clicked authorize - i ended up on 

https://localhost:5001/connect/authorize?client_id=b16f993c-9a3a-4ca8-b0fb-7db1c820e113&redirect_uri=https://localhost:5002/signin-oidc&response_type=code&scope=openid+profile+email&state=test&nonce=test

which requires auth working on this port... not sure if thats a problem but it looks like one :(

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Consent.cshtml.cs:
```diff
@@ -1,83 +1,141 @@
-        else
+
+        // Ensure the request contains the openid and profile scopes
+        RequestScopes = [..AuthRequest.GetScopes().Union(["openid", "profile"])];
```
shouldn't those be from constants?

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Consent.cshtml.cs:
```diff
@@ -1,83 +1,141 @@
+using OpenIddict.Server.AspNetCore;
 using SS14.Auth.Shared.Data;
+using SS14.Web.Helpers;
+using SS14.Web.Models.Types;
```
unused using

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Consent.cshtml.cs:
```diff
@@ -1,83 +1,141 @@
+        {
+            ConsentResult.ResultType.SignIn =>
+                SignIn(result.Principal!,
+                    authenticationScheme: OpenIddictServerAspNetCoreDefaults.AuthenticationScheme),
```
lets not linebreak it :Godo:

Fildrance commented on SS14.Web/Areas/Identity/Pages/Account/Manage/OAuthApps/Create.cshtml.cs:
```diff
@@ -1,43 +1,51 @@
 namespace SS14.Web.Areas.Identity.Pages.Account.Manage.OAuthApps;
 
 [ValidateAntiForgeryToken]
 public class Create : PageModel
```
Would be nice to remark key features for which user could get to admins for client settings customization (logo uri for example) on creation page.

Fildrance commented on SS14.Auth.Shared/Data/Migrations/20250812220730_SwitchToOpenIddict.cs:
```diff
@@ -0,0 +1,998 @@
+        /// <inheritdoc />
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            /*migrationBuilder.DropForeignKey(
```
i would say that it is useful to remark why it is commented. My guess is that you don't want tables to be dropped but you also don't want giant diff with those deletes in new migration in case you would need new one. Or, rather, i suppose you will delete it before merge?

Fildrance commented on SS14.Auth.Shared/Data/Migrations/20250815165542_RemoveSpaceAppSecretDescriptionProperty.cs:
```diff
@@ -0,0 +1,22 @@
+namespace SS14.Auth.Shared.Data.Migrations
+{
+    /// <inheritdoc />
+    public partial class RemoveSpaceAppSecretDescriptionProperty : Migration
```
huh

Fildrance commented on SS14.Auth.Shared/Data/Migrations/20250817184858_RemoveSecretDescriptionAndDate.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <inheritdoc />
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.DropColumn(
```
I think there is no point in splitting all those changes into different migrations? Can we just aggregate it? Or at least have basic migration that came up with default models of openiddict and one for additional stuff?

Fildrance commented on SS14.Web.Tests/TestMultipleClientSecretHandling.cs:
```diff
@@ -0,0 +1,44 @@
+
+public class TestMultipleClientSecretHandling
+{
+    [TestCase( "0.1755342104.key1111.desc", 0, ExpectedResult = 0)]
```
odd whitespaces at start of test case and unused usings

Fildrance commented on Tools/GenerateCerts.csx:
```diff
@@ -0,0 +1,35 @@
+
+certificate = request.CreateSelfSigned(DateTimeOffset.UtcNow, DateTimeOffset.UtcNow.AddYears(10));
+
+File.WriteAllBytes("server-signing-certificate-rsa-pss.pfx", certificate.Export(X509ContentType.Pfx, string.Empty));
```
I would say it would be nice to have something added in readme.md - what it does (what for certs should be used with openiddict in current scheme), when to use it.

Fildrance commented on Tools/identityserver4_to_openiddict_data_migration.sql:
```diff
@@ -0,0 +1,26 @@
+﻿insert into public."OpenIddictApplications"
+select gen_random_uuid() as Id,
+       (select "SpaceUserId" from public."UserOAuthClients" uc where c."Id" = uc."ClientId"),
+       "LogoUri", null as ApplicationType, c."ClientId",
+       (select string_agg(array_to_string(array [
+                                              cs."Id"::text,
+                                          floor(extract(epoch from cs."Created"))::text,
+                                          '_OLD_' || cs."Value",
+                                          replace(cs."Description", '*', '')
+                                              ], '.'), ',') from "IS4"."ClientSecrets" cs where cs."ClientId" = c."Id") ClientSecret,
+       'confidential' as ClientType, gen_random_uuid() as ConcurrencyToken,
+       case when "RequireConsent" then 'Explicit' else 'Implicit' end as "ConsentType",
+       "ClientName" as DisplayName, null as DisplayNames, null as JsonWebKeySet,
+       '["ept:authorization","ept:token","ept:introspection","ept:end_session","gt:refresh_token","gt:authorization_code","rst:code","scp:email","scp:profile","scp:roles"]'
+           as Permissions,
+       null as PostLogoutRedirectUris,
+       null as Properties,
+       (select json_agg("RedirectUri") from "IS4"."ClientRedirectUris" cr where cr."ClientId" = c."Id")::text "RedirectUris",
+    case when "RequirePkce" then '["ft:pkce"]' end as "Requirements",
+       json_build_object(
+           'space:AllowPlainPkce', "AllowPlainTextPkce"::TEXT,
+           'space:SigningAlgorithm', "AllowedIdentityTokenSigningAlgorithms",
+           'space:Disabled', ("Enabled" = false)::TEXT
+       )::text as "Settings",
+    "ClientUri" as WebsiteUrl
```
```suggestion
       (select "SpaceUserId" from public."UserOAuthClients" uc where c."Id" = uc."ClientId"),
       "LogoUri", 
       null as ApplicationType, 
       c."ClientId",
       (select string_agg(array_to_string(array [
                                              cs."Id"::text,
                                          floor(extract(epoch from cs."Created"))::text,
                                          '_OLD_' || cs."Value",
                                          replace(cs."Description", '*', '')
                                              ], '.'), ',') from "IS4"."ClientSecrets" cs where cs."ClientId" = c."Id") ClientSecret,
       'confidential' as ClientType, 
       gen_random_uuid() as ConcurrencyToken,
       case when "RequireConsent" then 'Explicit' else 'Implicit' end as "ConsentType",
       "ClientName" as DisplayName,
       null as DisplayNames, 
       null as JsonWebKeySet,
       '["ept:authorization","ept:token","ept:introspection","ept:end_session","gt:refresh_token","gt:authorization_code","rst:code","scp:email","scp:profile","scp:roles"]'
           as Permissions,
       null as PostLogoutRedirectUris,
       null as Properties,
       (select json_agg("RedirectUri") from "IS4"."ClientRedirectUris" cr where cr."ClientId" = c."Id")::text "RedirectUris",
       case when "RequirePkce" then '["ft:pkce"]' end as "Requirements",
       json_build_object(
           'space:AllowPlainPkce', "AllowPlainTextPkce"::TEXT,
           'space:SigningAlgorithm', "AllowedIdentityTokenSigningAlgorithms",
           'space:Disabled', ("Enabled" = false)::TEXT
       )::text as "Settings",
       "ClientUri" as WebsiteUrl
```

Fildrance replied:
fix indent and line breaks (1 field max per line)

Fildrance commented on .gitignore:
```diff
@@ -7,4 +7,5 @@ riderModule.iml
-*/tempkey.jwk
\ No newline at end of file
+*/tempkey.jwk
+cert
```
By default they are generated in tools folder, so i would propose ignoring them there and pathing them there :|

Fildrance commented on Tools/identityserver4_to_openiddict_data_migration.sql:
```diff
@@ -0,0 +1,26 @@
+﻿insert into public."OpenIddictApplications"
+select gen_random_uuid() as Id,
+       (select "SpaceUserId" from public."UserOAuthClients" uc where c."Id" = uc."ClientId"),
```
have we not deleted UserOAuthClients in  AddSpaceAppHomePageProperty migration?

Fildrance reviewed (changes requested):
Overall it works! Looks slim, which is nice. Almost all of my comments are kinda nitpicks and are dismissable, bragging about inconsistencies in codestyle (which is funny, considering we review stuff with razor pages, kek). I tried my best not to devolve into poking questionable linebreaks :>

Tested it with local calls, with OAuthTest, analyzed OAuthTest and it seems to be statifying all in all.

I do, however, propose some more additions before we release it:
* adding settings for blocking user/'oauth client' creation and updates (with some setting-provided message) - so we could bascially make database read-only and not have any data-loss in case of backup if something goes poorly during release
* remove tables deletion from migrations so old stuff would still work fine if we just revert app
* prepare migration up for migrating down
* hook migration script into c# migration class :>
